rust/tools/authorization-logic/src/souffle/lowering_ast_datalog.rs

```diff
@@ -0,0 +1,341 @@
+                // be a recursive call, so it might not be processing
+                // the top-level fact. The top-level fact gets wrapped
+                // in a "says" during the call to translate the assertion
+                // in which this appears.
+
+                // The following code generates
+                // p says fpf :- x says fpf, p says x canSay fpf
+                // where fpf is fact_plus_prime
+
+                // p says fact_plus_prime
+                let lhs = push_prin(String::from("says_"), p, &fact_plus_prime);
+                // x says fact_plus_prime
+                let x_says_term = push_prin(String::from("says_"), &x, &fact_plus_prime);
+                // p says x canSay fact_plus_prime
+                let can_say_term = push_prin(
+                    String::from("says_"),
+                    p,
+                    &push_prin(String::from("canSay_"), &x, &fact_plus_prime),
+                );
+                // p says fpf :- x says fpf, p says x canSay fpf
+                let mut rhs = Vec::new();
+                rhs.push(x_says_term);
+                rhs.push(can_say_term);
+                let gen = DLIRAssertion::DLIRCondAssertion { lhs, rhs };
+
+                collected.push(gen);
+                (f_prime, collected)
+            }
+        }
+    }
+
+    fn says_assertion_to_dlir(&mut self, x: &AstSaysAssertion) -> Vec<DLIRAssertion> {
+        x.assertions
+            .iter()
+            .map(|assertion| self.says_assertion_to_dlir_inner(&x.prin, assertion))
+            .flatten()
+            .collect()
+    }
+
+    fn says_assertion_to_dlir_inner(
+        &mut self,
+        speaker: &AstPrincipal,
+        assertion: &AstAssertion,
+    ) -> Vec<DLIRAssertion> {
+        match &assertion {
+            AstAssertion::AstFactAssertion { f } => {
+                let (pred, mut gen_assert) = self.fact_to_dlir(&f, &speaker);
+                let pred_prime = push_prin(String::from("says_"), &speaker, &pred);
+                gen_assert.push(DLIRAssertion::DLIRFactAssertion { p: pred_prime });
+                gen_assert
+            }
+            AstAssertion::AstCondAssertion { lhs, rhs } => {
+                let mut dlir_rhs = Vec::new();
+                for f in rhs {
+                    let (flat, _) = self.flat_fact_to_dlir(&f, &speaker);
+                    dlir_rhs.push(push_prin(String::from("says_"), &speaker, &flat));
+                }
+                let (lhs_prime, mut assertions) = self.fact_to_dlir(&lhs, &speaker);
+                let dlir_lhs = push_prin(String::from("says_"), &speaker, &lhs_prime);
+                let this_assertion = DLIRAssertion::DLIRCondAssertion {
+                    lhs: dlir_lhs,
+                    rhs: dlir_rhs,
+                };
+                assertions.push(this_assertion);
+                assertions
+            }
+        }
+    }
+
+    // This can't be a const because Strings (by contrast to &str's) can't be
```

Comment:
nit: applicable in multiple places: comment should end with a .

Reply:
Fixed.